File: src/main/java/com/netdetpla/ndp/controller/TaskController.java
package com.netdetpla.ndp.controller;

import com.netdetpla.ndp.bean.ResponseEnvelope;
import com.netdetpla.ndp.bean.SubTask;
import com.netdetpla.ndp.bean.Task;
import com.netdetpla.ndp.handler.DatabaseHandler;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.io.UnsupportedEncodingException;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;

import static com.netdetpla.ndp.utils.ImageUtil.*;
import static com.netdetpla.ndp.utils.IpUtil.*;

@RestController
public class TaskController {

    @PostMapping("/task")
    public ResponseEntity<?> addTask(
            @RequestParam("image-name") String imageName,
            @RequestParam("tag") String tag,
            @RequestParam("task-name") String taskName,
            @RequestParam("priority") String priority,
            @RequestParam("params[]") String params[]
    ) throws SQLException {
        System.out.println("image-name: " + imageName);
        System.out.println("tag: " + tag);
        System.out.println("task-name: " + taskName);
        System.out.println("priority: " + priority);
        for(int k=0;k<params.length;k++){
            System.out.println("param"+k+"：" + params[k]);
        }

        ResultSet resultSet = DatabaseHandler.executeQuery(
                "select id from image where image_name = ? and tag = ?",
                imageName,
                tag
        );
        resultSet.next();
        int image_id = resultSet.getInt(1);
        //获取id,tid
        ResultSet resultSet2 = DatabaseHandler.executeQuery(
                "select max(id),max(tid) from task");
        resultSet2.next();
        int id = resultSet2.getInt(1)+1;
        int tid = resultSet2.getInt(2)+1;
        String tidString = String.valueOf(tid);

        switch (imageName){
            case "scanweb":
                scanweb(id,tidString,image_id,taskName,priority,params,24);
                break;
            case "ecdsystem":
                ecdsystem(id,image_id,tidString,taskName,priority,params);
                break;
            case "scanservice":
                scanservice(id,tidString,image_id,taskName,priority,params,24);
                break;
            default:
                break;
        }


        return new ResponseEntity<>(new ResponseEnvelope<>(
                HttpStatus.OK.value(),
                "OK"
        ), HttpStatus.OK);
    }

    @GetMapping("/task/{image_name}/{task_name}")
    public ResponseEntity<?> getSubTask(@PathVariable(value = "task_name") String task_name) throws SQLException {
        List<SubTask> data = new ArrayList<>();
        ResultSet resultSet = DatabaseHandler.executeQuery(
                "select id,start_time,end_time,param,task_status,priority from task where task_name = ?",
                task_name
        );
        while (resultSet.next()) {
            data.add(new SubTask(
                    resultSet.getInt(1),
                    resultSet.getString(2),
                    resultSet.getString(3),
                    resultSet.getString(4),
                    resultSet.getString(5),
                    resultSet.getInt(6)
            ));
        }
        return new ResponseEntity<>(new ResponseEnvelope<>(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK);
    }

    @GetMapping("/task/{image_name}")
    public ResponseEntity<?> getTask(@PathVariable("image_name") String imageName) throws SQLException {
        // TODO 查询任务
        List<Task> data = new ArrayList<>();
        ResultSet resultSet = DatabaseHandler.executeQuery(
//                "select s.tid, s.task_name, s.start_time, s.end_time from " +
//                        "(select *, row_number() over (partition by tid order by id) as group_idx from task) s " +
//                        "where s.image_id in " +
//                        "(select id from image where image_name = ?) " +
//                        "and s.group_idx = 1",
                "select  tid, task_name, start_time, end_time from task where image_id in (select id from image where image_name = ?) GROUP BY tid ORDER BY tid desc",
                imageName
        );
        while (resultSet.next()) {
            data.add(new Task(
                    resultSet.getInt(1),
                    resultSet.getString(2),
                    resultSet.getString(3),
                    resultSet.getString(4)
            ));
        }
        return new ResponseEntity<>(new ResponseEnvelope<>(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK);
    }
}

File: src/main/kotlin/com/ndp/server/controller/TaskController.kt
package com.ndp.server.controller

import com.ndp.server.bean.ResponseEnvelope
import com.ndp.server.bean.TaskJson
import com.ndp.server.utils.DatabaseHandler
import com.ndp.server.utils.Miscellaneous
import com.ndp.server.utils.Settings
import com.ndp.server.utils.TaskGenerator
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
class TaskController {
    private val ipCount = Integer.parseInt(Settings.setting["ipCount"] as String)
    private val country = Settings.setting["ipCountry"] as String
    private val ports = Settings.setting["portScanPort"] as String
    private val dnsServer = Settings.setting["dnsServer"] as String

    @PostMapping("/task")
    fun addTask(
            @RequestParam("image-name") imageName: String,
            @RequestParam("tag") tag: String,
            @RequestParam("task-name") taskName: String,
            @RequestParam("priority") priority: String,
            @RequestParam("params[]") params: Array<String>
    ): ResponseEntity<*>? {
        println("image-name: $imageName")
        println("tag: $tag")
        println("task-name: $taskName")
        println("priority: $priority")
        for (k in params.indices) {
            println("param" + k + "：" + params[k])
        }

        val imageID = DatabaseHandler.selectImageID(imageName, tag)
        if (imageID == -1)
            return ResponseEntity(ResponseEnvelope<Any?>(
                    HttpStatus.BAD_REQUEST.value(),
                    "No such image & tag"
            ), HttpStatus.BAD_REQUEST)
        val tid = DatabaseHandler.selectMaxTid() + 1
        when (imageName) {
            "dnssecure" -> TaskGenerator.dnssecure(
                    tid,
                    imageID,
                    taskName,
                    priority,
                    params
            )
            "ip-test" -> TaskGenerator.ipTest(
                    tid,
                    imageID,
                    taskName,
                    priority,
                    params
            )
            "port-scan" -> TaskGenerator.portScan(
                    tid,
                    imageID,
                    taskName,
                    priority,
                    params
            )
            "page-crawl" -> TaskGenerator.pageCrawl(
                    tid,
                    imageID,
                    taskName,
                    priority,
                    params
            )
            "url-crawl" -> TaskGenerator.urlCrawl(
                    tid,
                    imageID,
                    taskName,
                    priority,
                    params
            )
        }
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Create task successfully."
        ), HttpStatus.OK)
    }

    @GetMapping("/task/running")
    fun getRunningTask(): ResponseEntity<*>? {
        val tids = DatabaseHandler.selectRunningTopTask()
        val data = ArrayList<TaskJson>()
        for (t in tids) {
            data.add(DatabaseHandler.selectTaskInfo(t))
        }
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/task/finished")
    fun getFinishedTask(): ResponseEntity<*>? {
        val data = Miscellaneous.parseChartMap(DatabaseHandler.getFinishedTask())
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/task/unfinished")
    fun getUnfinishedTask(): ResponseEntity<*>? {
        val data = Miscellaneous.parseChartMap(DatabaseHandler.getUnfinishedTask())
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/task/available/ipTestA")
    fun getIPTestAAvailable(): ResponseEntity<*>? {
        val data = DatabaseHandler.selectIPTestAAvailable()
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/task/available/portScan")
    fun getPortScanAvailable(): ResponseEntity<*>? {
        val data = DatabaseHandler.selectPortScanAvailable()
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/task/available/dnssecure")
    fun getDnssecureAvailable(): ResponseEntity<*>? {
        val data = DatabaseHandler.selectDnssecureAvailable()
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/task/available/urlCrawl")
    fun getURLCrawlAvailable(): ResponseEntity<*>? {
        val data = DatabaseHandler.selectURLCrawlAvailable()
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/task/available/pageCrawl")
    fun getPageCrawlAvailable(): ResponseEntity<*>? {
        val data = DatabaseHandler.selectPageCrawlAvailable()
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @PostMapping("/task/batchAdd/ipTestGeo")
    fun batchAddIPTestGeoTask(
            @RequestParam("num") num: String
    ): ResponseEntity<*>? {
        val limit = Integer.parseInt(num)
        val target = DatabaseHandler.getIPByCountry(country, ipCount, limit)
        Settings.setting["ipCount"] = (ipCount + limit).toString()
        Settings.save()
        val imageID = DatabaseHandler.selectNewImageID("ip-test")
        val tid = DatabaseHandler.selectMaxTid()
        for (t in target) {
            TaskGenerator.ipTest(
                    tid,
                    imageID,
                    "batch-add-ip-test-geo",
                    "5",
                    arrayOf(t)
            )
        }
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Create task successfully."
        ), HttpStatus.OK)
    }

    @PostMapping("/task/batchAdd/ipTestA")
    fun batchAddIPTestATask(
            @RequestParam("num") num: String
    ): ResponseEntity<*>? {
        val limit = Integer.parseInt(num)
        val target = DatabaseHandler.getIPByDomainIP(limit)
        val imageID = DatabaseHandler.selectNewImageID("ip-test")
        val tid = DatabaseHandler.selectMaxTid()
        for (t in target) {
            TaskGenerator.ipTest(
                    tid,
                    imageID,
                    "batch-add-ip-test-a",
                    "5",
                    arrayOf(t)
            )
        }
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Create task successfully."
        ), HttpStatus.OK)
    }

    @PostMapping("/task/batchAdd/portScan")
    fun batchAddPortScanTask(
            @RequestParam("num") num: String
    ): ResponseEntity<*>? {
        val limit = Integer.parseInt(num)
        val ips = DatabaseHandler.getIPByPortScanFlag(limit).joinToString(",")
        val imageID = DatabaseHandler.selectNewImageID("port-scan")
        val tid = DatabaseHandler.selectMaxTid()
        TaskGenerator.portScan(
                tid,
                imageID,
                "batch-add-port-scan",
                "5",
                arrayOf(ips, ports)
        )
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Create task successfully."
        ), HttpStatus.OK)
    }

    @PostMapping("/task/batchAdd/dnssecure")
    fun batchAddDnssecureTask(
            @RequestParam("num") num: String
    ): ResponseEntity<*>? {
        val limit = Integer.parseInt(num)
        val urls = DatabaseHandler.getUrlByDnssecureFlag(limit)
        val imageID = DatabaseHandler.selectNewImageID("dnssecure")
        val tid = DatabaseHandler.selectMaxTid()
        TaskGenerator.dnssecure(
                tid,
                imageID,
                "batch-add-dnssecure",
                "5",
                arrayOf(urls.joinToString("+"), dnsServer)
        )
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Create task successfully."
        ), HttpStatus.OK)
    }

    @PostMapping("/task/batchAdd/urlCrawl")
    fun batchAddURLCrawlTask(
            @RequestParam("num") num: String
    ): ResponseEntity<*>? {
        val limit = Integer.parseInt(num)
        val urls = DatabaseHandler.getUrlByUrlFlag(limit).joinToString(",")
        val imageID = DatabaseHandler.selectNewImageID("url-crawl")
        val tid = DatabaseHandler.selectMaxTid()
        TaskGenerator.urlCrawl(
                tid,
                imageID,
                "batch-add-url-crawl",
                "5",
                arrayOf(urls)
        )
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Create task successfully."
        ), HttpStatus.OK)
    }

    @PostMapping("/task/batchAdd/pageCrawl")
    fun batchAddPageCrawlTask(
            @RequestParam("num") num: String
    ): ResponseEntity<*>? {
        val limit = Integer.parseInt(num)
        val urls = DatabaseHandler.getUrlByUrlFlag(limit).joinToString(",")
        val imageID = DatabaseHandler.selectNewImageID("page-crawl")
        val tid = DatabaseHandler.selectMaxTid()
        TaskGenerator.pageCrawl(
                tid,
                imageID,
                "batch-add-page-crawl",
                "5",
                arrayOf(urls)
        )
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Create task successfully."
        ), HttpStatus.OK)
    }
}